torch/csrc/jit/codegen/cuda/lower_mem_index.cpp

```diff
@@ -0,0 +1,1222 @@
+      id = split->in();
+      id_def = id->definition();
+
+      // Already hitting contiguous root so stop
+      //  traversing up.
+      if (contig_merged_ids.count(id)) {
+        return true;
+      }
+
+      // Check that the split is divisible
+      if (!id->extent()->isConstInt()) {
+        return false;
+      }
+      auto id_extent = id->extent()->evaluateInt();
+      auto factor_extent = split->factor()->getInt();
+
+      if (!factor_extent.has_value()) {
+        // Non constant sized tiling cannot be lifted.
+        return false;
+      }
+
+      if (id_extent % factor_extent.value() != 0) {
+        // Non-divisible split cannot be separated.
+        return false;
+      }
+    } else if (auto merge = dynamic_cast<Merge*>(id_def)) {
+      // When we hit a merge, check for pattern 2
+
+      if (prev_id == nullptr) {
+        // A merge on the leaf domain not yet supported.
+        return false;
+      }
+
+      if (auto prev_split = dynamic_cast<Split*>(prev_id->definition())) {
+        auto split_leaves = getFlattenedSplitIds(id, {prev_id});
+        auto merged_leaves = getFlattenedMergedIds(id);
+
+        if (prev_id != split_leaves.at(0)) {
+          // Only support the case when both ids are leading dims.
+          return false;
+        }
+
+        // Check the actual static dimensions, which would be the
+        //  only way for now to prove that we'd not need to wrap
+        //  around when incrementing this id:
+        auto maybe_id_stride = getMaybeLeadingStride(merged_leaves);
+        auto maybe_prev_id_stride = getMaybeLeadingStride(split_leaves);
+
+        if (maybe_id_stride.has_value() && maybe_prev_id_stride.has_value()) {
+          if (maybe_id_stride.value() <= maybe_prev_id_stride.value()) {
+            if (!require_divisible ||
+                (maybe_prev_id_stride.value() % maybe_id_stride.value() == 0)) {
+              id = merged_leaves.at(0);
+              id_def = id->definition();
+              continue;
+            }
+          }
+        }
+      }
+
+      return false;
+    } else if (auto swizzle_2d = dynamic_cast<Swizzle2D*>(id_def)) {
+      if (ignore_swizzle) {
+        // Forward through swizzle if it is to be ignored.
+        id = id == swizzle_2d->outX() ? swizzle_2d->inX() : swizzle_2d->inY();
+        id_def = id->definition();
+      } else {
+        // When not ignored, swizzles are always assumed to be
+        //  not separable, i.e. they will take extra space in the pre-computed
+        //  tensor address space.
+        return false;
+      }
+    }
+  }
+
+  // This covers the base case where we start
+  //  with a contigous merged output of root.
+  return contig_merged_ids.count(id);
+}
+
+//! Traverse forward on the iterdomain's use chain and returns
+//!  true if a swizzle iterdomain op is found.
+bool isSwizzleProducer(IterDomain* id) {
+  auto single_use = ir_utils::getMaybeSingleUse(id, ir_utils::isIterDomainOp);
```

Comment:
Is there any specific reason to check about the single usage? As far as I remember, IterDomains are still used by only one expression, right?